File: src/test/java/net/igsoft/sdi/testclasses/CCreator.java
package net.igsoft.sdi.testclasses;

import java.util.List;

import com.google.common.collect.Lists;

import net.igsoft.sdi.creator.CreatorBase;
import net.igsoft.sdi.engine.InstanceProvider;
import net.igsoft.sdi.parameter.LaunchType;

public class CCreator extends CreatorBase<CClass, LaunchType> {
    @Override
    public CClass create(InstanceProvider instanceProvider, LaunchType launchType) {
        AClass a = instanceProvider.getOrCreate(AClass.class, launchType);
        BClass b = instanceProvider.getOrCreate(BClass.class, launchType);

        return new CClass(a, b, instanceProvider.getOrCreate(Stepper.class));
    }

    @Override
    public List<CreatorBase<?, ?>> defaultCreators() {
        return Lists.newArrayList(new ACreator(), new BCreator());
    }
}

File: src/test/scala/net/igsoft/sdi/scala/testclasses/CCreator.scala
package net.igsoft.sdi.scala.testclasses

import net.igsoft.sdi.{Creator, InstanceCreator, LaunchType}

class CCreator extends Creator[C, LaunchType] {
  override def create(instanceCreator: InstanceCreator, launchType: LaunchType): C = {
    val a = instanceCreator.getOrCreate(classOf[A], launchType)
    val b = instanceCreator.getOrCreate(classOf[B], launchType)
    new C(a, b, instanceCreator.getOrCreate(classOf[Stepper]))
  }
}
